### PR TITLE
ClusterLoader - Adding checker map

### PR DESCRIPTION
**Pull request by krzysied**

### CODE REVIEW

File: clusterloader2/pkg/measurement/util/checker/checker_map.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+type CheckerMap map[string]Checker
+
+// NewCheckerMap creates new checker map.
+func NewCheckerMap() CheckerMap {
+	return make(map[string]Checker)
+}
+
+// StopAllCheckers stops all checkers in the map.
+func (cm CheckerMap) StopAllCheckers() {
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/util/checker/checker_map.go]: Call it dispose and also clean up the map.

File: clusterloader2/pkg/measurement/util/checker/interface.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package checker
+
+type Checker interface {
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/util/checker/interface.go]: Add a comment.

File: clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go [hunk 3]
```diff
@@ -191,15 +190,16 @@ func (w *waitForControlledPodsRunningMeasurement) gather(syncTimeout time.Durati
 	var numberRunning, numberDeleted, numberTimeout, numberUnknown int
 	unknowStatusErrList := errors.NewErrorList()
 	timedOutObjects := []string{}
-	for _, checker := range w.checkerMap {
-		status, err := checker.getStatus()
+	for _, checker := range (map[string]checker.Checker)(w.checkerMap) {
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go]: Do you really need this cast?

File: clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go [hunk 4]
```diff
@@ -191,15 +190,16 @@ func (w *waitForControlledPodsRunningMeasurement) gather(syncTimeout time.Durati
...
+		objChecker, _ := (checker).(*objectChecker)
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go]: Don't silently ignore error.
If you're sure that the checker is of the type just do:

objCheck := checker.(*objectChecker)

File: clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go [hunk 5]
```diff
@@ -191,15 +190,16 @@ func (w *waitForControlledPodsRunningMeasurement) gather(syncTimeout time.Durati
...
+		status, err := objChecker.getStatus()
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go]: And you can combine those two lines then.

File: clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go [hunk 6]
```diff
@@ -317,9 +317,9 @@ func (w *waitForControlledPodsRunningMeasurement) handleObjectLocked(oldObj, new
 		operationTimeout *= 2
 	}
 	time.AfterFunc(operationTimeout, func() {
-		checker.terminate(true)
+		c.terminate(true)
 	})
-	w.checkerMap[key] = checker
+	((map[string]checker.Checker)(w.checkerMap))[key] = c
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go]: We don't want to do this.
If this case is really needed, you need to add a method on Checker map like SetChecker(key, checker)

File: clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go [hunk 7]
```diff
@@ -331,9 +331,10 @@ func (w *waitForControlledPodsRunningMeasurement) deleteObjectLocked(obj runtime
 	if err != nil {
 		return fmt.Errorf("meta key creation error: %v", err)
 	}
-	if checker, exists := w.checkerMap[key]; exists {
-		checker.terminate(false)
-		delete(w.checkerMap, key)
+	cm := (map[string]checker.Checker)(w.checkerMap)
+	if checker, exists := cm[key]; exists {
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go]: DeleteAndStopIf(key) bool
should be added to the interface and used here.

File: clusterloader2/pkg/measurement/util/checker/checker_map.go (hunk 1 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/util/checker/checker_map.go]: Done.

File: clusterloader2/pkg/measurement/util/checker/interface.go (hunk 2 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/util/checker/interface.go]: Done.

File: clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go (hunk 3 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go]: I thought that range cannot be applied to custom types. It turns out I was wrong.

File: clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go (hunk 4 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go]: Done.

File: clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go (hunk 5 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go]: objChecker is used three lines below so I need to keep the variable.

File: clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go (hunk 6 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go]: Done.

File: clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go (hunk 7 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go]: Done. Although I didn't add the return type.

File: clusterloader2/pkg/measurement/util/checker/interface.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (krzysied) [reply clusterloader2/pkg/measurement/util/checker/interface.go]: In most of the places we have a separate file for interfaces. Although it no consistent approach anymore :/
I'm not sure if there is a golang convention for this...
Assuming that the separate file is not always used I can merge those two...

File: clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go [hunk 8]
```diff
@@ -192,14 +191,15 @@ func (w *waitForControlledPodsRunningMeasurement) gather(syncTimeout time.Durati
 	unknowStatusErrList := errors.NewErrorList()
 	timedOutObjects := []string{}
 	for _, checker := range w.checkerMap {
-		status, err := checker.getStatus()
+		objChecker := (checker).(*objectChecker)
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go]: nit: checker.(*objectChecker)

[remove the () around checker]

File: clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go [hunk 9]
```diff
@@ -305,7 +305,7 @@ func (w *waitForControlledPodsRunningMeasurement) handleObjectLocked(oldObj, new
 	if err != nil {
 		return fmt.Errorf("meta key creation error: %v", err)
 	}
-	checker, err := w.waitForRuntimeObject(handledObj, isObjDeleted)
+	c, err := w.waitForRuntimeObject(handledObj, isObjDeleted)
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go]: please revert the name to checker - it's more intuitive

File: clusterloader2/pkg/measurement/util/checker/checker_map.go [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+// Add adds checker to the checker map.
+func (cm CheckerMap) Add(key string, c Checker) {
+	if old, exists := cm[key]; exists {
+		old.Stop()
+	}
+	cm[key] = c
+}
+
+// DeleteAndStop stops checker and deletes it if exists.
+func (cm CheckerMap) DeleteAndStop(key string) {
```
Comment (wojtek-t) [inline clusterloader2/pkg/measurement/util/checker/checker_map.go]: return bool whether it was deleted or stopped or not

[if clients don't needed, they can always ignore it]

File: clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go (hunk 8 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go]: Done.

File: clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go (hunk 9 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go]: Done.

File: clusterloader2/pkg/measurement/util/checker/checker_map.go (hunk 10 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/util/checker/checker_map.go]: Done.

**Pull request merged**